### PR TITLE
[Ads] Invibes ad integration

### PR DESCRIPTION
**Pull request by invibes**
- Invibes ad provider integration

✨ New feature (`:sparkles:`)

### CODE REVIEW

**Comment by invibes** [comment]
Confirm consent for pull request

**Comment by rcheptanariu** [comment]
Confirm consent for pull request

**Comment by raducheptanariu** [comment]
Hello
@zhouyx could you please review my code?

Is there something else required on my end?

Thanks !

File: ads/invibes.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/invibes.js]: nit: 2019

File: ads/invibes.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+<!---
+Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/invibes.md]: nit :2019

File: ads/invibes.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    data-pid="placement1"
+    data-ad-categ="infeed">
+</amp-ad>
+```
+
+## Configuration
+
+For configuration semantics, please contact [Invibes](https://www.invibes.com/#section-contact-email)
+
+Supported parameters:
```
Comment (zhouyx) [inline ads/invibes.md]: nit: Looks like those two are mandatory based on the implementation. Could we please label them as required parameters.

File: ads/invibes.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  };
+
+  validateData(
+    data,
+    global.invibesAmp.mandatoryData,
+    global.invibesAmp.allowedData
+  );
+
+  let url = 'https://k.r66net.com/GetAmpLink';
+  if (data.adCateg) {
```
Comment (zhouyx) [inline ads/invibes.js]: Given we've already verified `data.adCateg` in `#validateData()`, do we still need the if check here?

File: ads/invibes.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    url += '?adCateg=' + encodeURIComponent(data.adCateg);
+  }
+
+  if (data.pid) {
```
Comment (zhouyx) [inline ads/invibes.js]: ditto

File: ads/_config.js [hunk 6]
```diff
@@ -522,6 +522,12 @@ export const adConfig = {
     renderStartImplemented: true,
   },
 
+  'invibes': {
+    prefetch: 'https://k.r66net.com/GetAmpLink',
+    renderStartImplemented: true,
```
Comment (zhouyx) [inline ads/_config.js]: Thanks for implementing the renderStart API. When I was testing with the example you provide, I didn't see the render-start message. Could you please double check that the API is implemented correctly. Thank you

File: ads/_config.js [hunk 7]
```diff
@@ -522,6 +522,12 @@ export const adConfig = {
...
+    consentHandlingOverride: true,
```
Comment (zhouyx) [inline ads/_config.js]: I couldn't find the code, just to confirm that you have customized the ad behavior with the `getConsentState()` API within the iframe. Thanks

File: ads/invibes.md (hunk 2 above)
Comment (raducheptanariu) [reply ads/invibes.md]: fixed

File: ads/invibes.js (hunk 1 above)
Comment (raducheptanariu) [reply ads/invibes.js]: fixed

File: ads/invibes.md (hunk 3 above)
Comment (raducheptanariu) [reply ads/invibes.md]: Thanks for the remark! My implementation was faulty. We have a default value for those parameters on our backend so they should definitely be optional.

File: ads/invibes.js (hunk 4 above)
Comment (raducheptanariu) [reply ads/invibes.js]: There are no mandatory fields atm. We may choose to redirect to a different script based on the value of adCateg, or if the value is omitted. Do the additional changes to the code reflect this or is something else still required?

File: ads/invibes.js (hunk 5 above)
Comment (raducheptanariu) [reply ads/invibes.js]: There are no mandatory fields atm. We may choose to redirect to a different script based on the value of pid, or if the value is omitted (for example for value '1234' we give back a test script / creative) Do the additional changes to the code reflect this or is something else still required?

File: ads/_config.js (hunk 6 above)
Comment (raducheptanariu) [reply ads/_config.js]: Our GetAmpLink endpoint is not production ready. It's in testing and should be ready within the week. The example is with a test placement '1234' and only redirects to a basic script/creative which renders an image logo.
The script which will be deployed next week is at
http://static.videostepstage.com/scripts/amp.min.js

The renderStart is implemented like this
`
var _ampContext = window.context;

invibes.addEventListener('ivbsAdRendered', function () {
  _ampContext.renderStart();
);

invibes.addEventListener('ivbsFallback', function () {
  _ampContext.noContentAvailable();
});
`

The 'ivbsAdRendered' event is dispatched if we have an ad available from our ad server, certain context conditions are met etc..
Else the 'ivbsFallback' event is dispatched.

File: ads/_config.js (hunk 7 above)
Comment (raducheptanariu) [reply ads/_config.js]: Same as above, the GetAmpLink endpoint is not production ready.

```
var CONSENT_POLICY_STATE = {
  UNKNOWN: 0,
  SUFFICIENT: 1,
  INSUFFICIENT: 2,
  UNKNOWN_NOT_REQUIRED: 3
};

_ampContext.getConsentState(function (consent) {
  if (consent.consentState == CONSENT_POLICY_STATE.SUFFICIENT || consent.consentState == CONSENT_POLICY_STATE.UNKNOWN_NOT_REQUIRED) {
    consentManager.updateConsent(2);
  }
});
```

We consider updateConsent with value > 0 to be consent given and will show targeted ads. The default is consent = 0 and we will not give targeted ads.

**Comment by rcheptanariu** [comment]
Hello
/cc @zhouyx
@zhouyx have you had time to review my changes ? :)

File: ads/invibes.js (hunk 4 above)
Comment (zhouyx) [reply ads/invibes.js]: Thanks for the clarification. Since they are not mandatory data, this makes sense to me.

**Comment by zhouyx** [comment]
And to address the CLA comment
>😕 The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.

Could we please have all authors confirm that they are ok with their commits being contributed to the project? Thanks

File: ads/_config.js (hunk 7 above)
Comment (rcheptanariu) [reply ads/_config.js]: > @raducheptanariu Thank you for the clarification. Looks great to me. One question about the `GetAmpLink` endpoint. I tested it today, and it seems it's still not ready? Could you please double check.
> 
> It's not blocking the PR, but FYI we usually cut around every Tuesday. And the change will be live in prod one week after the pre-release cut. Thanks.


> @raducheptanariu Thank you for the clarification. Looks great to me. One question about the `GetAmpLink` endpoint. I tested it today, and it seems it's still not ready? Could you please double check.
> 
> It's not blocking the PR, but FYI we usually cut around every Tuesday. And the change will be live in prod one week after the pre-release cut. Thanks.

Hello and thanks for the feedback !

The endpoint for the client js is ready. The script is minified but you can see how the functions you referenced earlier are implemented (getConsentState, renderStart, noContentAvailable).

The ad server is not production ready though. This shouldn't be an issue though, as we require publishers be configured in our system prior to integration.

**Comment by rcheptanariu** [comment]
> And to address the CLA comment
> 
> > 😕 The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.
> 
> Could we please have all authors confirm that they are ok with their commits being contributed to the project? Thanks

All commits have been pushed from this account and I confirm they are OK being contributed to the project.

Also wrote in a comment after the initial pull request from both the pull request author (invibes) and the commit author (rcheptanariu) that we

>  Confirm consent for pull request

**Comment by zhouyx** [comment]
Thank you @rcheptanariu for the Pull Request. LGTM and Merged.

**Pull request merged**